### ISSUE TITLE
Resource Testding doesn't return a TextValueAsHtml

### ISSUE THREAD

**Issue by mdelez** [dasch-swiss/knora-api#1629]
Can we update this resource to return a TextValueAsHtml?

**Comment by benjamingeer** [comment]
Yes, will do.

**Comment by benjamingeer** [comment]
@tobiasschweizer If I understand correctly, the XSL transformation to HTML has to be specified in the TextValue. If I do that, the text value will always be returned as HTML, and it will no longer be possible to return it as XML. Is that correct? If so, I can't do that, because it would break existing tests.

**Comment by benjamingeer** [comment]
Or maybe I misunderstood. Is this a request to add an additional text value with an XSL transformation to Testding? Or is it a request to add a `textValueAsHTML` property to the existing text value?

**Comment by mdelez** [comment]
<img width="1061" alt="Screen Shot 2020-03-12 at 12 09 35" src="https://user-images.githubusercontent.com/60604010/76515822-9356e080-645a-11ea-941c-77c53c259167.png">

As of right now, the ReadResource object doesn't return a TextValueAsHtml property which I would like to have if possible :)

**Comment by benjamingeer** [comment]
@mdelez This isn't simple to do, because it can't be done at run time. To be able to serve a text value as HTML, you need to upload an XSL stylesheet to Sipi to do an XML-to-HTML transformation, and change the text value that is stored in the triplestore so that it refers to this stylesheet. It's not clear to me whether this is possible with our existing testing infrastructure.

@tobiasschweizer We have Knora-and-Sipi integration tests that create XSL transformations, but otherwise we don't have one in the test data at all. From the documentation, I'm having trouble seeing how to add an XSL transformation to the static `anything` test data.

**Comment by tobiasschweizer** [comment]
> We have Knora-and-Sipi integration tests that create XSL transformations, but otherwise we don't have one in the test data at all. From the documentation, I'm having trouble seeing how to add an XSL transformation to the static anything test data.

Tricky indeed. Could you manufacture a static case class that is turned into JSON-LD as you did for the static UUID test data?

**Comment by benjamingeer** [comment]
> Could you manufacture a static case class that is turned into JSON-LD as you did for the static UUID test data?

I could, but wouldn't it be better to use real Knora output? Otherwise the static case class could get out of sync with reality.

**Comment by tobiasschweizer** [comment]
> Otherwise the static case class could get out of sync with reality.

But if you generate the JSON-LD from a case class instance and it gets out of sync, then there would be a compile error, right?

**Comment by benjamingeer** [comment]
> But if you generate the JSON-LD from a case class instance and it gets out of sync, then there would be a compile error, right?

OK, now I see what you mean. I could make a hard-coded `ReadResourceV2`, containing a hard-coded `TextValueContentV2`, containing a hard-coded `MappingXMLtoStandoff`...

Do we have an existing XML-to-HTML mapping, or would I have to write one from scratch?

**Comment by tobiasschweizer** [comment]
> containing a hard-coded TextValueContentV2, containing a hard-coded MappingXMLtoStandoff...

Actually we do not care about the mapping. I think you can just invent some HTML, that's fine.

**Comment by benjamingeer** [comment]
> I think you can just invent some HTML, that's fine.

I can't do that if I have a hard-coded `TextValueContentV2` instance, because the HTML is generated in `TextValueContentV2.toJsonLDValue()`.

I could just use a hard-coded JSON-LD response (containing hard-coded HTML) in the test data, but then there would be no way to ensure that it stays consistent with Knora's real responses.

**Comment by tobiasschweizer** [comment]
> > I think you can just invent some HTML, that's fine.
> 
> I can't do that if I have a hard-coded `TextValueContentV2` instance, because the HTML is generated in `TextValueContentV2.toJsonLDValue()`.
> 
> I could just use a hard-coded JSON-LD response (containing hard-coded HTML) in the test data, but then there would be no way to ensure that it stays consistent with Knora's real responses.

I understand. So then could you make a very simple mapping and XSLT to generate some HTML?

**Comment by benjamingeer** [comment]
> So then could you make a very simple mapping and XSLT to generate some HTML?

Yes, that's why I'm asking... do we have one? Was there ever a test for this?

**Comment by tobiasschweizer** [comment]
I think I made something for this TEI route that involves some XSLT. 
There is surely some tests involving letters standoff to XML, but I am not sure if there is any XSLT.

**Comment by benjamingeer** [comment]
> There is surely some tests involving letters standoff to XML, but I am not sure if there is any XSLT.

That's in the Knora-Sipi integration tests. In `_test_data/test_route/texts/beol` there's `standoffToTEI.xsl` and `header.xsl`. The integration tests first load those files into Sipi, and then use them in the tests. Hence my original question: how can my client test data route do anything with an XSL tranformation? Does it have to first load the XSL file into Sipi? Or is there a way to include the XSLT in the static data for the `anything` project, so it's already there when the client test data route is called?

**Comment by tobiasschweizer** [comment]
> how can my client test data route do anything with an XSL tranformation? Does it have to first load the XSL file into Sipi? Or is there a way to include the XSLT in the static data for the anything project, so it's already there when the client test data route is called?

Normally, XSLT is a file that is managed by Sipi and requested and cached by Knora (very cumbersome to test for).

But in this case I think you could just provide some static data to `TextValueContentV2`: https://github.com/dasch-swiss/knora-api/blob/c41bce78750f2a565b5d994763327c78bb8c5125/webapi/src/main/scala/org/knora/webapi/messages/v2/responder/valuemessages/ValueMessagesV2.scala#L1286

**Comment by benjamingeer** [comment]
OK, I'll try it!

**Comment by tobiasschweizer** [comment]
> OK, I'll try it!

Good luck!

**Comment by benjamingeer** [comment]
Since the standard mapping already returns HTML tags (`<p>`, etc.), would it be possible to use an XSL transformation that does nothing, and just returns the XML as-is?

**Comment by tobiasschweizer** [comment]
yes, I think you can apply an identity transformation: https://en.wikipedia.org/wiki/Identity_transform#Using_XSLT

It just copies what you have.

**Comment by benjamingeer** [comment]
@tobiasschweizer 

> But in this case I think you could just provide some static data to TextValueContentV2

I've just realised that I can't do this, because the test data generation just does an HTTP GET to a Knora route. It doesn't have access to the `TextValueContentV2`.

I think it might be necessary to redesign this feature so that it's possible to test it.

**Comment by tobiasschweizer** [comment]
> @tobiasschweizer
> 
> > But in this case I think you could just provide some static data to TextValueContentV2
> 
> I've just realised that I can't do this, because the test data generation just does an HTTP GET to a Knora route. It doesn't have access to the `TextValueContentV2`.
> 
> I think it might be necessary to redesign this feature so that it's possible to test it.

I can finish https://github.com/dasch-swiss/knora-ui-ng-lib/pull/55 with some static test data and will just replace it once it's possible. So this is not a big problem for me.

**Comment by benjamingeer** [comment]
> So this is not a big problem for me.

I think it’s a problem to have a feature that isn’t tested and can’t be tested.

**Comment by tobiasschweizer** [comment]
> > So this is not a big problem for me.
> 
> I think it’s a problem to have a feature that isn’t tested and can’t be tested.

It's a problem that there isn't JSON test data in the knora-api-js-lib so we can check for the correct deserialization, yes. But we can simulate it in the test data for knora-ui.

**Comment by benjamingeer** [comment]
What I mean is that Knora has a feature which we can’t test as part of Knora’s own tests. That doesn’t seem good to me.

**Comment by tobiasschweizer** [comment]
> What I mean is that Knora has a feature which we can’t test as part of Knora’s own tests. That doesn’t seem good to me.

No, it certainly isn't. We should refactor it. I can help.
